File: analysis/revision/scripts/incipient.py
import pandas as pd, numpy as np
import matplotlib; matplotlib.use('Agg'); import matplotlib.pyplot as plt
df=pd.read_csv('./data/Tree_ID_info.csv'); df['site']=np.where(df['plot']=='BGS','BGS','EMS')
sm={'rm':'A.rubrum','bg':'N.sylvatica','ro':'Q.rubra','hem':'T.canadensis'}; df['sp']=df['species'].map(sm)
ert=pd.read_csv('./data/ERT_application_results.csv'); m=df.merge(ert,on='tree').reset_index(drop=True)
# species-standardized resistivity deviation; ORIENT so higher = WETTER (right)
m['xdev']=-m.groupby('sp')['mean'].transform(lambda x:(x-x.median())/x.std(ddof=1))
m['struct']=m['percent_damaged']>1
sound=~m['struct']; dmgd=m['struct']; mb=(m.site=='BGS').values; me=(m.site=='EMS').values

# ---- WHERE DOES INCIPIENT START? sensitivity of the sound->incipient cut ----
print("Incipient onset = sound tree confidently WETTER than its species median.")
print("(xdev = species-standardized; wetter = higher; +0.5 SD = confidently wet)")
print(f"{'cut (SD wetter than median)':32s}{'nII':>5}{'BGS%':>7}{'EMS%':>7}")
for cut,lab in [(0,'median (any wetter)'),(0.25,'+0.25 SD'),(0.5,'+0.5 SD'),(1.0,'+1.0 SD')]:
    inc=sound&(m['xdev']>cut)
    b=inc[mb].sum()/sound[mb].sum()*100; e=inc[me].sum()/sound[me].sum()*100
    print(f"  {lab:30s}{inc.sum():>5}{b:>7.0f}{e:>7.0f}")

# ---- internal 'positive control': moisture of structurally-confirmed decay (active) trees ----
# active = damaged & wet (xdev>0). Their xdev shows 'decay-associated moisture' range.
active=m[dmgd & (m['xdev']>0)]
print(f"\nInternal reference — structurally-damaged WET (active) trees, n={len(active)}:")
print(f"  their xdev (wetness): median={active['xdev'].median():.2f}, range {active['xdev'].min():.2f}..{active['xdev'].max():.2f}")
print(f"  -> a sound tree wetter than ~+0.5 SD overlaps the active-decay moisture range (supports +0.5 SD onset)")

# ---- 6-cell / dead-band classification ----
def classify(cut):
    wet=m['xdev']>cut; dry=m['xdev']<-cut
    cat=np.where(sound & wet,'II',
        np.where(sound & ~wet,'I',
        np.where(dmgd & dry,'IV','III')))
    return cat
for cut in [0.5,1.0]:
    from collections import Counter
    print(f"\n6-cell scheme, dead band ±{cut} SD:  {dict(Counter(classify(cut)))}")

# ---- FIGURE: 6-cell dead-band phase diagram (two dead-band widths) ----
catcol={'I':'#3b6fb0','II':'#4e9a2c','III':'#d98a1f','IV':'#b83232'}
np.random.seed(2)
fig,axes=plt.subplots(1,2,figsize=(15,6),sharey=True)
for ax,cut in zip(axes,[0.5,1.0]):
    cat=classify(cut); y=np.sqrt(np.clip(m['percent_damaged'],0,None))
    ax.axvspan(-cut,cut,color='0.85',alpha=0.6,zorder=0)
    ax.axvline(0,color='#666',ls='-',lw=1); ax.axhline(np.sqrt(1),color='#666',ls='-',lw=1)
    for c,col in catcol.items():
        idx=cat==c
        ax.scatter(m['xdev'][idx], y[idx]+np.random.normal(0,0.03,idx.sum()), c=col, s=48, edgecolor='white',lw=.5,zorder=3,label=c)
    yt=[0,1,5,10,20,30]; ax.set_yticks([np.sqrt(t) for t in yt]); ax.set_yticklabels(yt)
    ax.set_xlabel('moisture anomaly  (species-standardized;  drier ←   0 = species median   → wetter)',fontsize=9)
    # cell labels
    xl=ax.get_xlim()
    ax.text(xl[0]+0.2,np.sqrt(31),'IV Cavity\n(damaged, confidently dry)',fontsize=8,color=catcol['IV'],weight='bold',va='top')
    ax.text(xl[1]-0.2,np.sqrt(31),'III Active\n(damaged, wet)',fontsize=8,color=catcol['III'],weight='bold',ha='right',va='top')
    ax.text(xl[0]+0.2,0.05,'I No decay\n(sound, not-wet)',fontsize=8,color=catcol['I'],weight='bold')
    ax.text(xl[1]-0.2,0.05,'II Incipient\n(sound, confidently wet)',fontsize=8,color=catcol['II'],weight='bold',ha='right')
    ax.text(0,np.sqrt(34),f'dead band ±{cut} SD\n(transition → defaults to row baseline)',fontsize=7.5,ha='center',color='#555')
    ax.set_title(f'Dead band ±{cut} SD',fontsize=10)
axes[0].set_ylabel('SoT structural loss %  (√ scale)')
axes[0].legend(title='Category',fontsize=8.5,frameon=False,loc='center left')
fig.suptitle('6-cell scheme: 4 confident corners + a "normal" transition column that defaults to the row baseline\n'
             '(sound→No-decay, damaged→Active). Only "confidently wet sound" (II) and "confidently dry damaged" (IV) are off-baseline.',
             fontsize=11,y=1.02)
plt.tight_layout(); plt.savefig('analysis/revision/output/CJFR-6cell.png',dpi=140,bbox_inches='tight')
print("\nsaved 6-cell figure")

File: analysis/revision/scripts/incipient.R
# Where does 'incipient' start? Sensitivity of the sound->incipient cut and a
# 6-cell dead-band scheme. R port of incipient.py. Run from the repo root:
#   Rscript analysis/revision/scripts/incipient.R

source("analysis/revision/scripts/revision_common.R")

# UTF-8 locale so the png devices render em-dash / sqrt / arrow glyphs
if (!grepl("UTF-8", Sys.getlocale("LC_CTYPE"))) invisible(try(Sys.setlocale("LC_CTYPE", "en_US.UTF-8"), silent = TRUE))

m <- load_merged()
# species-standardized resistivity deviation; ORIENT so higher = WETTER (right)
m$xdev <- xdev_species_median(m)
m$struct <- m$percent_damaged > 1
sound <- !m$struct; dmgd <- m$struct
mb <- m$site == "BGS"; me <- m$site == "EMS"

# ---- WHERE DOES INCIPIENT START? sensitivity of the sound->incipient cut ----
cat("Incipient onset = sound tree confidently WETTER than its species median.\n")
cat("(xdev = species-standardized; wetter = higher; +0.5 SD = confidently wet)\n")
cat(sprintf("%-32s%5s%7s%7s\n", "cut (SD wetter than median)", "nII", "BGS%", "EMS%"))
cuts <- list(c(0, "median (any wetter)"), c(0.25, "+0.25 SD"),
             c(0.5, "+0.5 SD"), c(1.0, "+1.0 SD"))
for (cl in cuts) {
  cut_ <- as.numeric(cl[1]); lab <- cl[2]
  inc <- sound & (m$xdev > cut_)
  b <- sum(inc[mb]) / sum(sound[mb]) * 100
  e <- sum(inc[me]) / sum(sound[me]) * 100
  cat(sprintf("  %-30s%5d%7.0f%7.0f\n", lab, sum(inc), b, e))
}

# ---- internal 'positive control': moisture of structurally-confirmed decay (active) trees ----
# active = damaged & wet (xdev>0). Their xdev shows 'decay-associated moisture' range.
active <- m[dmgd & (m$xdev > 0), ]
cat(sprintf("\nInternal reference — structurally-damaged WET (active) trees, n=%d:\n", nrow(active)))
cat(sprintf("  their xdev (wetness): median=%.2f, range %.2f..%.2f\n",
            median(active$xdev), min(active$xdev), max(active$xdev)))
cat("  -> a sound tree wetter than ~+0.5 SD overlaps the active-decay moisture range (supports +0.5 SD onset)\n")

# ---- 6-cell / dead-band classification ----
classify <- function(cut_) {
  wet <- m$xdev > cut_; dry <- m$xdev < -cut_
  ifelse(sound & wet, "II",
    ifelse(sound & !wet, "I",
      ifelse(dmgd & dry, "IV", "III")))
}
cat_counts <- function(cc) {
  t <- table(factor(cc, levels = c("I", "II", "III", "IV")))
  paste(sprintf("%s=%d", names(t), as.integer(t)), collapse = " ")
}
for (cut_ in c(0.5, 1.0)) {
  cat(sprintf("\n6-cell scheme, dead band ±%s SD:  {%s}\n",
              sprintf("%.1f", cut_), cat_counts(classify(cut_))))
}

# ---- FIGURE: dead-band phase diagram, nested band widths in one panel ----
# Rationale for the widths (PC1 sd = 1.89): the alternative data-driven
# thresholds sit at -0.13 SD (median), +0.53 SD (Otsu) and +1.03 SD (GMM
# crossover) from the mean, so +/-0.5 SD spans the median-to-Otsu range of
# defensible cuts and +/-1 SD additionally covers the GMM crossover.
set.seed(2)
yt <- c(0, 1, 5, 10, 20, 30)
y <- sqrt(pmax(m$percent_damaged, 0))
xr <- range(m$xdev); xpad <- 0.05 * diff(xr)
xlim <- c(xr[1] - xpad, xr[2] + xpad)

png(file.path(OUT_DIR, "CJFR-6cell.png"),
    width = 8.4, height = 5.6, units = "in", res = 170)
par(mar = c(4.4, 4.2, 1.2, 0.8), mgp = c(2.4, 0.7, 0))
cat_ <- classify(0)   # baseline species-median split; bands overlay uncertainty
plot(NA, xlim = xlim, ylim = c(-0.2, sqrt(37) + 0.2), axes = FALSE,
     xlab = "", ylab = "")
usr <- par("usr")
rect(-1.0, usr[3], 1.0, usr[4], col = adjustcolor("gray82", alpha.f = 0.45), border = NA)
rect(-0.5, usr[3], 0.5, usr[4], col = adjustcolor("gray62", alpha.f = 0.45), border = NA)
abline(v = 0, col = "#666666", lty = 1, lwd = 1)
abline(h = sqrt(1), col = "#666666", lty = 1, lwd = 1)
for (cc in names(CAT_COLS)) {
  idx <- cat_ == cc
  points(m$xdev[idx], y[idx] + rnorm(sum(idx), 0, 0.03),
         pch = 21, bg = CAT_COLS[[cc]], col = "white", lwd = 0.5, cex = 1.3)
}
axis(1, cex.axis = 0.8)
axis(2, at = sqrt(yt), labels = yt, cex.axis = 0.8, las = 1)
box()
title(xlab = "moisture anomaly  (species-standardized;  drier \u2190   0 = species median   \u2192 wetter)",
      cex.lab = 0.8)
title(ylab = "SoT structural loss %  (\u221a scale)", cex.lab = 0.85)
xl <- xlim
text(xl[1] + 0.2, sqrt(31), "IV Cavity", cex = 0.75,
     col = CAT_COLS[["IV"]], font = 2, adj = c(0, 1))
text(xl[2] - 0.2, sqrt(31), "III Active", cex = 0.75,
     col = CAT_COLS[["III"]], font = 2, adj = c(1, 1))
text(xl[1] + 0.2, sqrt(2.6), "I No decay", cex = 0.75,
     col = CAT_COLS[["I"]], font = 2, adj = c(0, 0.5))
text(xl[2] - 0.2, sqrt(2.6), "II Incipient", cex = 0.75,
     col = CAT_COLS[["II"]], font = 2, adj = c(1, 0.5))
text(0, sqrt(35.5), "transitional \u00b10.5 SD", cex = 0.62, col = "#444444")
text(0.75, sqrt(33), "\u00b11 SD", cex = 0.62, col = "#777777", adj = 0)
legend("left", legend = names(CAT_COLS), pch = 21, pt.bg = unname(CAT_COLS),
       col = "white", pt.cex = 1.25, cex = 0.75, bty = "n", title = "Category")
invisible(dev.off())
cat("\nsaved 6-cell figure\n")
